File: c++/347_Top_K_Frequent_Elements.cpp
#include <queue>
#include <unordered_map>
#include <vector>

using namespace std;

class Top_K_Frequent_Elements {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        vector<int> result;
        unordered_map<int, int> counts;
        priority_queue<pair<int,int>> maxHeap;

        for (int &num:nums){
            counts[num]++;
        }

        for (auto num:counts){
            maxHeap.push({num.second,num.first});
        }
        
        for (int i = 0; i < k; i++){
            result.push_back(maxHeap.top().second);
            maxHeap.pop();
        }
        
        return result;
    }
};

File: python/347_Top_K_Frequent_Elements.py
class Solution:
  def topKFrequent(self, nums: list[int], k: int) -> list[int]:
    counts = {}
    freq = [[] for i in range(len(nums) + 1)]
    for num in nums:
      counts[num] = counts.get(num, 0) + 1
    for c, n in counts.items():
      freq[c].append(n)
    res = []
    for i in range(len(freq) - 1, 0, -1):
      for n in freq[i]:
        res.append(n)
        if len(res) == k:
          return res